source/Nevermore/Util/DataModificationQueryBuilder.cs

```diff
@@ -330,87 +331,158 @@ CommandParameterValues GetDocumentParameters(Func<DocumentMap, string> allocateI
+                            parameters.Add(relatedVariableName, relatedDocumentId);
+                            if (x > 0)
+                                insertStatementBuilder.Append(",");
+                            insertStatementBuilder.AppendLine(
+                                $"(@{parentIdVariable}, '{mapping.TableName}', @{relatedVariableName}, '{relatedTableName}')");
+                            remaining--;
+
+                            if (parameters.Count >= SqlServerParameterLimit)
+                            {
+                                reachedParameterLimit = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    commands.Add(new PreparedCommand(insertStatementBuilder.ToString(), parameters, RetriableOperation.Insert, mapping, options.CommandTimeout));
+                    if (remaining > 0)
+                    {
+                        insertStatementBuilder = new StringBuilder();
+                        parameters = new CommandParameterValues();
+                        parameters.AddRange(documentParametersCreator());
+                    }
                 }
             }
+
+            return commands.ToArray();
         }
 
-        string AppendRelatedDocumentStatementsForUpdate(
-            string statement,
-            CommandParameterValues parameters,
+        PreparedCommand[] CreateUpdateCommandsWithRelatedDocuments(
+            string updateStatement,
             DocumentMap mapping,
-            object document)
+            object document,
+            UpdateOptions updateOptions)
         {
+            Func<CommandParameterValues> parametersCreator = () => GetDocumentParameters(
+                m => throw new Exception("Cannot update a document if it does not have an ID"),
+                null,
+                null,
+                document,
+                mapping,
+                DataModification.Update
+            );
+            var commands = new List<PreparedCommand>();
             var relatedDocumentData = GetRelatedDocumentTableData(mapping, new[] {document});
+            var parameters = parametersCreator();
+
             if (relatedDocumentData.Count == 0)
-                return statement;
+            {
+                commands.Add(new PreparedCommand(updateStatement, parameters, RetriableOperation.Update, mapping, updateOptions.CommandTimeout));
+                return commands.ToArray();
+            }
 
             var sb = new StringBuilder();
-            sb.AppendLine(statement);
+            sb.AppendLine(updateStatement);
             sb.AppendLine();
 
+            // sql commands are executed inside sp_executesql which means table variables and local temp tables won't live between commands
+            var globalTempTableName = $"##references_{Guid.NewGuid().ToString().Replace('-', '_')}";
```

Comment:
This is the best solution I could come up with to allow this approach to work over multiple commands.

Global temp tables are effectively reference counted and only exist while there are sessions open with references to the global temp table, so cleanup is automatically taken care of.